# Tarea 2

## Parte 1: Carga y Exploración Inicial

* Cargar los Datos: Carga el archivo CSV desde la URL proporcionada en un DataFrame de Pandas.

Inspección Inicial:
* Muestra las primeras 5 filas del DataFrame usando .head().
* Utiliza .info() para obtener un resumen de las columnas y sus tipos de datos.
* En una celda de Markdown, escribe una breve observación sobre los datos (ej: ¿Cuántos registros hay?, ¿los tipos de datos son correctos?).

In [1]:
import pandas as pd

url = 'https://raw.githubusercontent.com/aiplanethub/Datasets/refs/heads/master/Amazon%20Top%2050%20Bestselling%20Books%202009%20-%202019.csv'
df = pd.read_csv(url)
df.head()

Name  \
0                      10-Day Green Smoothie Cleanse   
1                                  11/22/63: A Novel   
2            12 Rules for Life: An Antidote to Chaos   
3                             1984 (Signet Classics)   
4  5,000 Awesome Facts (About Everything!) (Natio...   

                     Author  User Rating  Reviews  Price  Year        Genre  
0                  JJ Smith          4.7    17350      8  2016  Non Fiction  
1              Stephen King          4.6     2052     22  2011      Fiction  
2        Jordan B. Peterson          4.7    18979     15  2018  Non Fiction  
3             George Orwell          4.7    21424      6  2017      Fiction  
4  National Geographic Kids          4.8     7665     12  2019  Non Fiction

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         550 non-null    object 
 1   Author       550 non-null    object 
 2   User Rating  550 non-null    float64
 3   Reviews      550 non-null    int64  
 4   Price        550 non-null    int64  
 5   Year         550 non-null    int64  
 6   Genre        550 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 30.2+ KB


### Observaciones sobre los datos

- El DataFrame contiene **550 registros** y **7 columnas**.
- Las columnas son: Name, Author, User Rating, Reviews, Price, Year y Genre.
- Los tipos de datos son apropiados: texto para los nombres y autores, numéricos para calificaciones, reseñas, precio y año.
- No hay valores nulos en ninguna columna.
- El dataset incluye información sobre los 50 libros más vendidos de Amazon entre 2009 y 2019, con detalles como autor, género, calificación de usuario, número de reseñas, precio y año.

## Parte 2: Análisis con Programación Funcional

1. Uso de map para Crear una Nueva Columna:

In [9]:
# Crear una nueva columna 'Rating Label' usando map
def get_rating_label(rating):
    if rating >= 4.5:
        return 'Excellent'
    elif rating >= 4.0:
        return 'Good'
    elif rating >= 3.5:
        return 'Average'
    else:
        return 'Poor'

# Aplicar la función usando map
df['Rating Label'] = list(map(get_rating_label, df['User Rating']))

# Mostrar las primeras filas para verificar
print("Nueva columna 'Rating Label' creada:")
df[['Name', 'User Rating', 'Rating Label']].head(5)

Nueva columna 'Rating Label' creada:


Name  User Rating Rating Label
0                      10-Day Green Smoothie Cleanse          4.7    Excellent
1                                  11/22/63: A Novel          4.6    Excellent
2            12 Rules for Life: An Antidote to Chaos          4.7    Excellent
3                             1984 (Signet Classics)          4.7    Excellent
4  5,000 Awesome Facts (About Everything!) (Natio...          4.8    Excellent

In [8]:
# Usar map con lambda para crear etiquetas de calificación
df['Rating Label'] = list(map(lambda rating: "Excelente" 
if rating >= 4.7 
else "Bueno", df['User Rating']))

# Mostrar las primeras filas para verificar
print("Nueva columna 'Rating Label' creada con lambda:")
df[['Name', 'User Rating', 'Rating Label']].head(5)

Nueva columna 'Rating Label' creada con lambda:


Name  User Rating Rating Label
0                      10-Day Green Smoothie Cleanse          4.7    Excelente
1                                  11/22/63: A Novel          4.6        Bueno
2            12 Rules for Life: An Antidote to Chaos          4.7    Excelente
3                             1984 (Signet Classics)          4.7    Excelente
4  5,000 Awesome Facts (About Everything!) (Natio...          4.8    Excelente

2. Uso de filter para Seleccionar Libros de Élite

In [11]:
# Crear lista de precios usando filter para libros de no ficción con rating >= 4.8
precios_libros_elite = list(filter(
    lambda x: x['Genre'] == 'Non Fiction' and x['User Rating'] >= 4.8,
    df.to_dict('records')
))

# Extraer solo los precios de la lista filtrada
precios = [libro['Price'] for libro in precios_libros_elite]

print(f"Número de libros que cumplen las condiciones: {len(precios)}")
print(f"Precios de los libros de no ficción con rating >= 4.8: {precios}")
#print(f"Precio promedio: {sum(precios)/len(precios):.2f}")

Número de libros que cumplen las condiciones: 62
Precios de los libros de no ficción con rating >= 4.8: [12, 13, 24, 11, 11, 15, 18, 4, 5, 10, 7, 54, 11, 4, 22, 15, 15, 17, 7, 12, 8, 8, 8, 8, 8, 8, 6, 8, 15, 16, 27, 16, 11, 22, 20, 20, 6, 6, 16, 8, 8, 8, 8, 8, 12, 12, 12, 12, 0, 13, 5, 17, 14, 21, 16, 16, 16, 13, 16, 11, 4, 14]


In [12]:
# Crear lista de precios usando filter para libros de no ficción con rating >= 4.8
precios_libros_elite = list(filter(
    lambda x: x['Genre'] == 'Non Fiction' and x['User Rating'] >= 4.8,
    df.to_dict('records')
))

# Extraer solo los precios de la lista filtrada
precios = [libro['Price'] for libro in precios_libros_elite]

print(f"Número de libros que cumplen las condiciones: {len(precios)}")
print(f"Precios de los libros de no ficción con rating >= 4.8: {precios}")

Número de libros que cumplen las condiciones: 62
Precios de los libros de no ficción con rating >= 4.8: [12, 13, 24, 11, 11, 15, 18, 4, 5, 10, 7, 54, 11, 4, 22, 15, 15, 17, 7, 12, 8, 8, 8, 8, 8, 8, 6, 8, 15, 16, 27, 16, 11, 22, 20, 20, 6, 6, 16, 8, 8, 8, 8, 8, 12, 12, 12, 12, 0, 13, 5, 17, 14, 21, 16, 16, 16, 13, 16, 11, 4, 14]


## Parte 3: Aplicando Programación Orientada a Objetos (POO)

1. Crear la Clase Book:



In [13]:
class Book:
    def __init__(self, name, author, price, rating):
        self.name = name
        self.author = author
        self.price = price
        self.rating = rating
    
    def display_info(self):
        return f'Título: "{self.name}", Autor: {self.author}, Precio: ${self.price}'
    
    def is_highly_rated(self, threshold=4.7):
        return self.rating >= threshold

# Ejemplo de uso
libro_ejemplo = Book("1984", "George Orwell", 8, 4.7)
print(libro_ejemplo.display_info())
print(f"¿Es altamente calificado? {libro_ejemplo.is_highly_rated()}")

Título: "1984", Autor: George Orwell, Precio: $8
¿Es altamente calificado? True


In [16]:
# Elegir el primer libro del DataFrame
primer_libro = df.iloc[6]
libro_objeto = Book(
    name=primer_libro['Name'],
    author=primer_libro['Author'],
    price=primer_libro['Price'],
    rating=primer_libro['User Rating']
)

print("Libro seleccionado del DataFrame:")
print(libro_objeto.display_info())
print(f"¿Es altamente calificado? {libro_objeto.is_highly_rated()}")
print(f"Rating: {libro_objeto.rating}")

Libro seleccionado del DataFrame:
Título: "A Game of Thrones / A Clash of Kings / A Storm of Swords / A Feast of Crows / A Dance with Dragons", Autor: George R. R. Martin, Precio: $30
¿Es altamente calificado? True
Rating: 4.7
